crates/gpui/src/elements/text.rs

```diff
@@ -305,11 +305,53 @@ impl TextLayout {
                 }
 
                 let mut line_wrapper = cx.text_system().line_wrapper(text_style.font(), font_size);
-                let text = if let Some(truncate_width) = truncate_width {
-                    line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis)
+                let (text, truncated) = if let Some(truncate_width) = truncate_width {
+                    (
+                        line_wrapper.truncate_line(text.clone(), truncate_width, ellipsis),
+                        true,
+                    )
                 } else {
-                    text.clone()
+                    (text.clone(), false)
                 };
+                if truncated {
+                    //----------------------------------------------
+                    // Case 0: Normal
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 12, ... }
+                    //
+                    // Truncate res: abcd… (truncate_at = 4)
+                    // Run res: Run0 { string: abcd…, len: 7, ... }
+                    // ---------------------------------------------
+                    // Case 1: Drop some runs
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 4, ... }, Run1 { len: 4, ... }, Run2 { len: 4, ... }
+                    //
+                    // Truncate res: abcdef… (truncate_at = 6)
+                    // Runs res: Run0 { string: abcd, len: 4, ... }, Run1 { string: ef…, len:
+                    // 5, ... }
+                    // ----------------------------------------------
+                    // Case 2: Truncate at start of some run
+                    // Text: abcdefghijkl
+                    // Runs: Run0 { len: 4, ... }, Run1 { len: 4, ... }, Run2 { len: 4, ... }
+                    //
+                    // Truncate res: abcdefgh… (truncate_at = 8)
+                    // Runs res: Run0 { string: abcd, len: 4, ... }, Run1 { string: efgh, len:
+                    // 4, ... }, Run2 { string: …, len: 3, ... }
```

Comment:
Rather than have all of this as a comment, can we pull this change out into a function and then add a unit test in this file for each of these cases?

Reply:
[text.zip](https://github.com/user-attachments/files/17197986/text.zip)

@mikayla-maki Use this, I changed it by the way and added a test here。

On Windows, I don't know how to get a git patch, so just edit file online and override this file.

Reply:
<img width="623" alt="image" src="https://github.com/user-attachments/assets/e8831c51-6c77-4ec4-98c6-f8a6b4fc410a">

Reply:
I think you need to make a PR against @JunkuiZhang's repository, or pull her commits into a separate branch and add yours on top, and then make your own PR :)

Reply:
I'm going to use a different approach here.